`routes` macro allowing multiple paths per handler

ModProg opened the pull request:
<!-- Thanks for considering contributing actix! -->
<!-- Please fill out the following to get your PR reviewed quicker. -->

## PR Type
<!-- What kind of change does this PR make? -->
<!-- Bug Fix / Feature / Refactor / Code Style / Other -->
Feature


## PR Checklist
<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt.
- [x] (Team) Label with affected crates and semver status.


## Overview
<!-- Describe the current and new behavior. -->
<!-- Emphasize any breaking changes. -->
Adds support for `routes` macro, allowing multiple paths and methods to be specified.


<!-- If this PR fixes or closes an issue, reference it here. -->
<!-- Closes #000 -->
Closes #2713

ModProg committed:
WIP: basic implementation for `routes` macro

ModProg committed:
chore: changelog, docs, tests

ModProg commented:
This is now working and documented (already using it my self in https://github.com/ModProg/paste).

Would like some feedback on the code to get this in a merge able state ASAP, so I don't need to run on a git version forever.

popzxc commented on actix-web-codegen/tests/trybuild.rs:
```diff
@@ -12,6 +12,8 @@ fn compile_macros() {
     t.compile_fail("tests/trybuild/route-unexpected-method-fail.rs");
     t.compile_fail("tests/trybuild/route-malformed-path-fail.rs");
 
+    t.pass("tests/trybuild/routes-ok.rs");
```
Maybe add some negative tests? E.g. with `routes` macro without any actual routes supplied?

ModProg replied:
I think in that case it would just do nothing, but an error would make sense.

popzxc reviewed (approved):
LGTM

ModProg committed:
error on missing methods

popzxc commented on actix-web-codegen/src/lib.rs:
````diff
@@ -104,6 +115,39 @@ pub fn route(args: TokenStream, input: TokenStream) -> TokenStream {
+/// ```
+///
+/// # Attributes
+/// The `routes` macro it self has no parameters, but allows specifying the attribute macros for
````
```suggestion
/// The `routes` macro itself has no parameters, but allows specifying the attribute macros for
```

popzxc commented on actix-web-codegen/src/lib.rs:
```diff
@@ -104,6 +115,39 @@ pub fn route(args: TokenStream, input: TokenStream) -> TokenStream {
+///
+/// # Attributes
+/// The `routes` macro it self has no parameters, but allows specifying the attribute macros for
+/// the different methods, e.g. [`GET`](macro@get) or [`POST`](macro@post).
```
```suggestion
/// the multiple paths and/or methods, e.g. [`GET`](macro@get) and [`POST`](macro@post).
```

popzxc reviewed (approved):
Nice!

ModProg committed:
Apply suggestions from code review

Co-authored-by: Igor Aleksanov <popzxc@yandex.ru>

ModProg commented:
Is there anything left for me to do to make this PR mergeable?

robjtede commented:
I haven't reviewed yet but I'm targetting this for v4.2, not the upcoming v4.1 so it will be merged after that release.

robjtede committed:
Merge branch 'master' into advanced_route_macro

robjtede committed:
Merge branch 'master' into advanced_route_macro

robjtede committed:
Merge branch 'master' into advanced_route_macro

robjtede committed:
update test stderr expectation

robjtede committed:
add additional tests

robjtede committed:
fix stderr output

robjtede committed:
remove useless ResourceType

this is dead code from back when .to and .to_async were different ways to add a service

robjtede commented:
CI issue was just missing a feature flag on `syn`. Cleaned up a bit too. Looking good now.

Thanks for the contribution.

ModProg commented:
Thanks, really looking forward to 4.2 now :)
